tower-layer/src/layer_fn.rs

```diff
@@ -0,0 +1,25 @@
+use super::Layer;
+
+/// Returns a new `impl Layer` with the given closure.
+pub fn layer_fn<F, S, T>(f: F) -> impl Layer<S, Service = T> + Clone
+where
+    F: Fn(S) -> T + Clone,
+{
+    LayerFn { f }
+}
+
+#[derive(Clone, Copy, Debug)]
```

Comment:
As an aside, I'm not sure if the `Debug` impl is valuable here, even if we weren't returning `impl Layer`. AFAIK, no type implementing `Fn` is `Debug`, and user code can't implement the `Fn` trait for its own types, so I don't think the derived impl will ever work.

It might be better to provide a `Debug` impl that doesn't depend on the function being `Debug`. For example, we could do something like this:

```rust 
impl<F> fmt::Debug for LayerFn<F> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("LayerFn")
            .field("f", &format_args!("<{}>", std::any::type_name::<F>())
            .finish()
    }
}
``` 
(but, this would also require us to return a concrete, named type...)

Reply:
Could we return `impl Layer<_> + Clone + fmt::Debug`?

Reply:
I'm fine with exposing the named type, but I don't think there's a whole lot of value in preserving `Copy`... `Clone` should be Just Fine imo

Reply:
Doesn't sound like there is much gained by using `impl Layer` but potential for loosing out on stuff. Let me know if I should revert to `LayerFn` 😊

If so I'll also add the `Debug` impl like you suggested @hawkw. I always add `#[derive(Debug)]` out of habit.

Reply:
I think it's probably a good idea to have a `Debug` that works, or else anything holding this layer also can't easily be `Debug`. I think it should be possible to return `impl Layer<_> + Clone + fmt::Debug`, though, as @olix0r suggested.

I don't care too strongly about whether or not we return a named type from this. I'm just not really sure if `impl Layer<_>` has any actual _benefit_. It doesn't make this code a whole lot more concise, since we still have to have a type implementing `Layer`, it just lets us not make it be `pub`.

Reply:
Hehe okay. I've reverted to exposing and returning `LayerFn` and added the debug impl suggested.